keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,427 @@
+initial interval then note that the maximum service time is much
+larger than the initial interval; by contrast, in networking, the
+maximum service time (i.e., packet length / link speed) is much
+smaller than the initial interval.  Even if we take 1 minute as our
+initial interval, we still do not have the sort of relationship that
+obtains in networking.  Note that in order to get good statistics on a
+queue --- which is needed by the CoDel technique --- there have to be
+many requests served during an interval.  Because of this mismatch,
+and because equivalence of context has not been established, we are
+not agreed that the CoDel technique can be used.
+
+Note that the resource limit being applied is a distinct concept from
+the fairness criteria.  For example, in CPU scheduling there may be 4
+CPUs and 50 threads being scheduled onto those CPUs; we do not suppose
+the goal is to have each thread to be using 0.08 CPUs at each instant;
+a thread uses eitehr 0 or 1 CPUs at a given instant.  Similarly, in
+networking, a router may multiplex a thousand flows onto one link; the
+goal is not to have each flow use 1/1000th of the link at each
+instant; a packet uses 0 links while queued and 1 link while being
+transmitted.  Each CPU or link is used for just one thing at a time;
+this is the resource limit.  The fairness goal is about utilization
+observed over time.  So it is in our scenario too.  For example, we
+may have 5000 flows of requests and a concurrency limit of 600
+requests at any one time.  That does not mean that our goal is for
+each flow to have 0.12 requests running at each instant.  Our goal is
+to limit the number of running requests to 600 at each instant and
+provide some fairness in utilization averaged over time.
+
+That average over time must not be over too much or too little time.
+It would not make sense to average over all past time; that would
+allow a flow to build up a huge amount of credit, enabling it to crowd
+out other flows.  It also does not make sense for the average to cover
+a small amount of time.  Because serving requests, like transmitting
+packets, is lumpy we must average over many service times.  Approaches
+to this include: using a sequence of intervals, using a sliding
+window, and using an exponential decay.
+
+Another open issue is the categorization: what are the categories and
+how is a request assigned to a category?  We seem to be agreed on at
+least one important point: each request is assigned to exactly one
+category, and is handled by exactly one "bucket" or "queue".  We also
+seem to be converging toward a two-level hierarchy of categories,
+aligned with the handling outline discussed earlier: at the higher
+level there are priorities, and within each priority level there is a
+collection of flows that compete fairly.
+
+One question is whether the prioritization is strict, or there is some
+sort of leakage between priority levels.
```

Comment:
I think leakage is problematic because it weakens the semantics of
priority.  Pracitally, leakage may weaken that semantic to a large
degree.  Leakage may be motivated by thinking that some low priority
requests should be handled instead of high priority requests in some
circumstances.  Such circumstances should be clearly identified.  If
the thinking is that priority inversions are impossible to avoid then
that should be explained as such.